src/proto.rs

```diff
@@ -340,6 +341,95 @@ pub struct AllowCredentials {
+#[repr(u8)]
+pub enum CredentialProtectionPolicy {
+    /// This reflects "FIDO_2_0" semantics. In this configuration, performing
+    /// some form of user verification is optional with or without credentialID
+    /// list. This is the default state of the credential if the extension is
+    /// not specified.
+    UserVerificationOptional = 0x1,
+    /// In this configuration, credential is discovered only when its
+    /// credentialID is provided by the platform or when some form of user
+    /// verification is performed.
+    UserVerificationOptionalWithCredentialIDList = 0x2,
+    /// This reflects that discovery and usage of the credential MUST be
+    /// preceded by some form of user verification.
+    UserVerificationRequired = 0x3,
+}
+
+impl TryFrom<u8> for CredentialProtectionPolicy {
+    type Error = &'static str;
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        use CredentialProtectionPolicy::*;
+        match v {
+            0x1 => Ok(UserVerificationOptional),
+            0x2 => Ok(UserVerificationOptionalWithCredentialIDList),
+            0x3 => Ok(UserVerificationRequired),
+            _ => Err("Invalid policy number"),
+        }
+    }
+}
+
+impl From<CredentialProtectionPolicy> for u8 {
+    fn from(policy: CredentialProtectionPolicy) -> Self {
+        policy as u8
+    }
+}
+
+impl TryFrom<u8> for CredProtectExtensionClientOutput {
+    type Error = <CredentialProtectionPolicy as TryFrom<u8>>::Error;
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        CredentialProtectionPolicy::try_from(v)
+            .map(|policy| CredProtectExtensionClientOutput(policy))
+    }
+}
+
+impl From<CredProtectExtensionClientOutput> for u8 {
+    fn from(policy: CredProtectExtensionClientOutput) -> Self {
+        u8::from(policy.0)
+    }
+}
+
+/// Wrapper struct to (de)serialize [CredentialProtectionPolicy] as a number
+#[derive(Debug, Serialize, Clone, Deserialize)]
+#[serde(try_from = "u8", into = "u8")]
+pub struct CredProtectExtensionClientOutput(CredentialProtectionPolicy);
+
+/// The input for the credProtect webauthn extension
+/// https://fidoalliance.org/specs/fido-v2.1-rd-20210309/fido-client-to-authenticator-protocol-v2.1-rd-20210309.html#sctn-credProtect-extension
+#[derive(Debug, Serialize, Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CredProtectExtensionClientInput {
+    /// The credential policy to enact
+    pub credential_protection_policy: CredentialProtectionPolicy,
+    /// Whether it is better for the authenticator to fail to create a
+    /// credential rather than ignore the protection policy
+    /// If no value is provided, the client treats it as `false`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub enforce_credential_protection_policy: Option<bool>,
+}
+
+/// Extension option inputs for [PublicKeyCredentialCreationOptions]
+#[derive(Debug, Serialize, Clone, Deserialize)]
+pub struct AuthenticatorExtensionsClientInputs {
```

Comment:
This type will likely need a AuthenticatorExtensionsBuilder to make it easier for people to create this in their implementations.